src/compiler/core.ts

```diff
@@ -2482,6 +2482,12 @@ namespace ts {
         VeryAggressive = 3,
     }
 
+    /**
+     * Safer version of `Function` which should not be called.
+     * Every function should be assignable to this, but this should not be assignable to every function.
+     */
+    export type AnyFunction = (...args: never[]) => {} | null | undefined | void;

```

Comment:
why not use something more specific in each location, e.g. in an input position, either `(...args: any[]): T` if you care about input (e.g. `memoize`) or `(...args: any[]): void` if you do not care about the output  (e.g. `assert`)?

Reply:
I think this is more specific as it can't be called (with arguments at least) and the return value can't be used without casting. If we use `...args: any[]` we can pass in a function taking a `number` but then call it on a `string` and there will be no warning at compile-time.

Reply:
most of these are not called, e.g. getFunctionName, assert,, fail, etc.. and we never call them with `null` or `void`.. and the ones that accept `undefined` are usually marked with optional parameters.. 
for these it seems more correct to jut have `(...args: any[])=>void` or `(...args: never[]) => void` if you want to be strict.